### PR TITLE
Rewrite amp-fx-collection code to allow new presets

### PR DESCRIPTION
**Pull request by nainar**
This PR rewrites the amp-fx-collection related code to:
* Move all the boiler plate logic related to the FxProviders into `fx-provider.js`
* Add all preset specific logic into `amp-fx-presets.js`

This serves the following benefits:
* We don't instantiate classes incurring the performance cost
* Makes it easier to add new presets without boiler plate.

### CODE REVIEW

**Comment by nainar** [comment]
@alanorozco does this read more like the idea you had suggested?

File: extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+      // User provided factor is 1-based for easier understanding.
+      // Also negating number since we are using tranformY so negative = upward,
+      // positive = downward.
+      const adjustedFactor = -(parseFloat(fxElement.factor_) - 1);
+      const top = entry.positionRect.top;
+      // Offset is how much extra to move the element which is position within
+      // viewport times adjusted factor.
+      const offset = (fxElement.adjustedViewportHeight_ - top) * adjustedFactor;
+      fxElement.offset_ = offset;
```
Comment (alanorozco) [inline extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: Can we expose an interface for `offset` and `element`? It makes me a little nervous that we're modifying private members.

LGTM except for that.

Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: You're absolutely right. Made the change in the latest commit.

**Comment by nainar** [comment]
@alanorozco I made the change to use accessors. Could you merge this in when you get the chance? Thanks!

File: extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      const adjustedFactor = -(parseFloat(fxElement.getFactor()) - 1);
+      const top = entry.positionRect.top;
+      // Offset is how much extra to move the element which is position within
+      // viewport times adjusted factor.
+      const offset = (fxElement.adjustedViewportHeight_ - top) * adjustedFactor;
+      fxElement.setOffset(offset);
+
+      if (!fxElement.isMutateScheduled()) {
+        fxElement.setIsMutateScheduled(true);
+        fxElement.resources_.mutateElement(fxElement.getElement(), function() {
```
Comment (alanorozco) [inline extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: `getResources()` or `mutateElement()` directly.

Comment (nainar) [reply extensions/amp-fx-collection/0.1/providers/amp-fx-presets.js]: Done!

**Pull request merged**